### ISSUE TITLE
how to hook flutter pub get and meanwhile exec a script

### ISSUE THREAD

**Issue by allanGoldZhiKang** [flutter/flutter#138417]
### Is there an existing issue for this?

- [X] I have searched the [existing infra issues](https://github.com/flutter/flutter/issues?q=is%3Aopen+is%3Aissue+label%3Ateam-infra)

### Type of Request

bug

### Infrastructure Environment

LUCI, Github, Cocoon scheduler, Autosubmit, etc...

### What is happening?

how to hook flutter pub get and meanwhile exec a script

### Steps to reproduce

Step 1:
Step 2:
..
Step n:

### Expected results

I expect to see X when Y is finished.

**Comment by ricardoamador** [comment]
Adding a note that this is not an infra issue but seems to be some sort of help request or implementation help request.

**Comment by huycozy** [comment]
Hi @allanGoldZhiKang 
Could you be more specific about the issue? (the issue context, what did you try, the problem you are facing on, etc)

**Comment by allanGoldZhiKang** [comment]
@huycozy 
For example, I'm going to run a script to add a git hook in the.git directory to standardize git commits before I do flutter pub get

**Comment by huycozy** [comment]
@allanGoldZhiKang Are you having a problem when doing that? If it's not a problem, is it a proposal or feature request? 

In case it's not a bug or feature request to Flutter framework, please also see flutter's [community](https://flutter.dev/community) for resources and ask questions like this. Thank you!

**Comment by allanGoldZhiKang** [comment]
@huycozy 
It would be fine if I could add some configuration to the yaml file so that I could execute a script at the same time as or before the flutter pub get.
But no matter how hard I tried, I couldn't achieve it. Can I just say that this is a design flaw caused by flutter's use of yaml files?
Can you teach me how to make flutter yaml like cocoapods, add some ruby configuration to podfile to execute some scripts and so on

thank you very much

**Comment by huycozy** [comment]
Looks like this is suited to tool team. I will label this for other's input.

**Comment by bkonyi** [comment]
@jonasfj @sigurdm, I'm guessing we aren't interested in adding hooks for pre/post `pub get` based on https://github.com/dart-lang/pub/issues/2190?

**Comment by sigurdm** [comment]
> I'm guessing we aren't interested in adding hooks for pre/post pub get

Yeah that is true. We want the `pub get` operation to be runnable without arbitrary side-effects.
One kind of hooks we **might** want to allow is [pre-publish hooks](https://github.com/dart-lang/pub/issues/4058) (for validating the package), but I don't think that is what this issue asks for. 

> For example, I'm going to run a script to add a git hook in the.git directory to standardize git commits 

Not sure I understand this. What exact command would you like to run, and why run it on pub get?

Have you looked into [git pre-commit hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks#_email_hooks:~:text=Enforced%20Policy.-,Committing%2DWorkflow%20Hooks,-The%20first%20four)?

**Comment by allanGoldZhiKang** [comment]
> > I'm guessing we aren't interested in adding hooks for pre/post pub get
> 
> Yeah that is true. We want the `pub get` operation to be runnable without arbitrary side-effects. One kind of hooks we **might** want to allow is [pre-publish hooks](https://github.com/dart-lang/pub/issues/4058) (for validating the package), but I don't think that is what this issue asks for.
> 
> > For example, I'm going to run a script to add a git hook in the.git directory to standardize git commits
> 
> Not sure I understand this. What exact command would you like to run, and why run it on pub get?
> 
> Have you looked into [git pre-commit hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks#_email_hooks:~:text=Enforced%20Policy.-,Committing%2DWorkflow%20Hooks,-The%20first%20four)?

do you know package.json and husky:

//package.json
{
  "husky":{
    "hooks":{
      "pre-commit":"npm test && npm run foo"
    }
  }
}

do you mean yaml couldn't do it？ @sigurdm

**Comment by sigurdm** [comment]
>  do you mean yaml couldn't do it

No I'm saying we don't want to do it.

**Comment by allanGoldZhiKang** [comment]
> > do you mean yaml couldn't do it
> 
> No I'm saying we don't want to do it.

i sorry to hear that
Why is frontend JavaScript actively addressing pre-hook hooks, while Flutter seems reluctant to implement these effective and useful features?

**Comment by sigurdm** [comment]
> i sorry to hear that
Why is frontend JavaScript actively addressing pre-hook hooks, while Flutter seems reluctant to implement these effective and useful features?

These issues have been discussed here: https://github.com/dart-lang/pub/issues/2190

Basically it boils down to (in my opinion): post-get is the wrong time to run code.

You want to be able to fetch dependencies in a predictable manner - if you allow all packages to run code after they are installed that becomes difficult.

What you most often want is some sort of build-step that can set up the packages according to the current target platform (that is unknown at pub-get time)

I can also repeat my questions:  

> What exact command would you like to run, and why run it on pub get?

> Have you looked into [git pre-commit hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks#_email_hooks:~:text=Enforced%20Policy.-,Committing%2DWorkflow%20Hooks,-The%20first%20four)?

**Comment by sigurdm** [comment]
```
//package.json
{
  "husky":{
    "hooks":{
      "pre-commit":"npm test && npm run foo"
    }
  }
}
```

Looking at this example again I realize this is a pre commit hook, not a post get hook.

I wonder why that is not just set up with a git pre-commit hook? Why do you need pub to support this?

**Comment by jonasfj** [comment]
quick brain dump, nothing fully thought through here:

We could also consider a post-pub-get hook that only works in the root package.
Meaning the post-pub-get hook in your dependencies won't get called.

Or we could imagine something where the root `pubspec.yaml` contains a list of dependencies that are authorized to run post-get-hooks.

But these options aren't as ergonomic as what `npm` post-install hooks can do. On the other hand, `npm`-style post-install hooks can be a bit of a security nightmare.

The other problem is that such hooks will be used to drive build systems / code-gen, etc. which will work great at first. But probably could make incremental builds harder to do, and make hot-reload work less well.

There is on going work on native assets, which might cover some of these uses, because I suspect they'll essentially do something that'll be hooked into the build system, and support incremental builds.

-----

I am curious though: What are the use-cases? can we start with that.

**Comment by sigurdm** [comment]
There is nothing in pub preventing you from writing a package like husky (https://typicode.github.io/husky/get-started.html) that you could add as a dev-dependency `dart pub add dev:dart_husky`.

Then add a `dart_husky` key to pubspec.yaml:
```
dart_husky:
  commit: "some shell command"
```

And then to `dart run dart_husky init` which would look for hooks in pubspec.yaml under the `dart_husky` key and install it as a commit hook in the current ./git folder.

I don't think we need post-install hooks for that.
